File: src/main/kotlin/sunsetsatellite/vm/sunlite/SLClassInstance.kt
package sunsetsatellite.vm.sunlite

import sunsetsatellite.lang.sunlite.Type

class SLClassInstance(val clazz: SLClass, val typeParams: MutableMap<String, Type>, val fields: MutableMap<String, SLField>) {

	override fun toString(): String {
		return "<object '${clazz.name}'>"
	}

	fun copy(): SLClassInstance {
		return SLClassInstance(clazz.copy(), typeParams.toMutableMap(),fields.mapValues { it.value.copy() }.toMutableMap())
	}
}

File: src/main/csharp/SunliteSharp/Runtime/SLClassInstance.cs
namespace SunliteSharp.Runtime;

using System.Collections.Generic;
using System.Linq;
using Type = SunliteSharp.Core.Type;

public record SLClassInstance(
    SLClass Clazz,
    Dictionary<string, Type> TypeParams,
    Dictionary<string, SLField> Fields
)
{
    public override string ToString() => $"<object '{Clazz.Name}'>";

    public SLClassInstance Copy()
    {
        return new SLClassInstance(
            Clazz.Copy(),
            new Dictionary<string, Type>(TypeParams),
            Fields.ToDictionary(kv => kv.Key, kv => kv.Value.Copy())
        );
    }
        
}